packages/belos/tpetra/test/MINRES/test_minres_diag.cpp

```diff
@@ -0,0 +1,418 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+//
+// This test generates diagonal matrices for MINRES to solve.
+//
+// NOTE: No preconditioner is used in this case.
+//
+
+// Belos includes
+#include "BelosConfigDefs.hpp"
+#include "BelosLinearProblem.hpp"
+#include "BelosTpetraAdapter.hpp"
+#include "BelosTpetraOperator.hpp"
+#include "BelosMinresSolMgr.hpp"
+
+// Tpetra includes
+#include <Tpetra_Core.hpp>
+#include <Tpetra_Map.hpp>
+#include <Tpetra_MultiVector.hpp>
+#include <Tpetra_Vector.hpp>
+
+// Teuchos includes
+#include <Teuchos_Comm.hpp>
+#include <Teuchos_ParameterList.hpp>
+#include <Teuchos_Time.hpp>
+#include <Teuchos_GlobalMPISession.hpp>
+#include <Teuchos_StandardCatchMacros.hpp>
+
+//
+/* TD: add typedefs as it is the common way in tpetra tests */
+// Get default Tpetra template types
+typedef Tpetra::MultiVector<>::scalar_type ST;
+typedef Tpetra::Vector<>::local_ordinal_type LO;
+typedef Tpetra::Vector<>::global_ordinal_type GO;
+typedef Tpetra::Vector<>::node_type NT;
+// Init Tpetra types
+typedef Tpetra::Operator<ST,LO,GO,NT> OP;
+typedef Tpetra::MultiVector<ST,LO,GO,NT> MV;
+typedef Tpetra::Map<LO,GO,NT> MP;
+
+//************************************************************************************************
+
+class Vector_Operator
+{
+  public:
+
+    Vector_Operator(int m_in, int n_in) : m(m_in), n(n_in) {};
+
+    virtual ~Vector_Operator() {};
+
+    virtual void operator () (const MV &x, MV &y) = 0;
+
+    int size (int dim) const { return (dim == 1) ? m : n; };
+
+  protected:
+
+    int m, n;        // an (m x n) operator
+
+  private:
+
+    // Not allowing copy construction.
+    Vector_Operator( const Vector_Operator& ): m(0), n(0) {};
+    Vector_Operator* operator=( const Vector_Operator& ) { return NULL; };
+
+};
+
+//************************************************************************************************
+
+class Diagonal_Operator : public Vector_Operator
+{
+  public:
+
+    Diagonal_Operator(int n_in, double v_in) : Vector_Operator(n_in, n_in), v(v_in) { };
+
+    ~Diagonal_Operator() { };
+
+    void operator () (const MV &x, MV &y)
+    {
+      y.scale( v, x );
+    };
+
+  private:
+
+    double v;
+};
+
+//************************************************************************************************
+
+class Diagonal_Operator_2 : public Vector_Operator
+{
+  public:
+
+    Diagonal_Operator_2(int n_in, int min_gid_in, double v_in)
+    : Vector_Operator(n_in, n_in), min_gid(min_gid_in), v(v_in) {}
+
+    ~Diagonal_Operator_2() { };
+
+    void operator () (const MV &x, MV &y)
+    {
+        auto yLocalData = y.getLocalViewHost(Tpetra::Access::ReadWrite);
+        auto xLocalData = y.getLocalViewHost(Tpetra::Access::ReadOnly);
+
+        for (size_t i = 0; i < x.getLocalLength(); ++i) {
```

Comment:
Epetra have:
- `MyLength()` which returns the local vector length
- `GlobalLength()` which returns the global vector length
- `GlobalLength64()` which returns the 64-bit global vector length

And they can be different so we have to be aware of this!